src/mapping/bd_rhapsody/script.py

```diff
@@ -0,0 +1,407 @@
+      'feature_type': feature_type,
+      'feature_id': [line[1:].strip() for line in f if line[0] == ">"],
+      'reference_file': os.path.basename(path),
+    })
+    return df
+
+def process_gtf(feature_type: str, path: str) -> pd.DataFrame:
+  with open(path) as f:
+    data = []
+    for line in f:
+      if not line.startswith("#"):
+        attr = dict(item.strip().split(' ') for item in line.split('\t')[8].strip('\n').split(';') if item)
+        row = {
+          'feature_type': feature_type,
+          'feature_id': attr["gene_name"].strip("\""),
+          'reference_file': os.path.basename(path),
+        }
+        data.append(row)
+  df = pd.DataFrame(data)
+  df = df.drop_duplicates()
+  return df
+
+def extract_feature_types(par: dict[str, Any]):
+  feature_types = pd.DataFrame()
+
+  if par["mode"] == "targeted":
+    for file in par["reference"]:
+      logger.info(f"Processing reference fasta {file}")
+      ft = process_fasta("Gene Expression", file)
+      feature_types = pd.concat([feature_types, ft])
+
+  if par["mode"] == "wta":
+    file = par["transcriptome_annotation"]
+    logger.info(f"Processing reference gtf {file}")
+    ft = process_gtf("Gene Expression", file)
+    feature_types = pd.concat([feature_types, ft])
+
+  if par["abseq_reference"]:
+    for file in par["abseq_reference"]:
+      logger.info(f"Processing abseq fasta {file}")
+      ft = process_fasta("Antibody Capture", file)
+      feature_types = pd.concat([feature_types, ft])
+
+  if par["supplemental_reference"]:
+    for file in par["supplemental_reference"]:
+      logger.info(f"Processing supp fasta {file}")
+      ft = process_fasta("Other", file)
+      feature_types = pd.concat([feature_types, ft])
+  
+  return feature_types
+
+def main(par: dict[str, Any], meta: dict[str, Any]):
+  # Preprocess params
+  par = process_params(par)
+
+  # Create output dir if not exists
+  if not os.path.exists(par["output"]):
+    os.makedirs(par["output"])
+
+  # Create params file
+  config_file = os.path.join(par["output"], "config.yml")
+  config_content = generate_config(par)
+  with open(config_file, "w") as f:
+    f.write(config_content)
+
+  # Create cwl file (if need be)
+  cwl_file = generate_cwl_file(par, meta)
+
+  ## Process parameters
+  proc_pars = ["--no-container", "--outdir", par["output"]]
+
+  if par["parallel"]:
+    proc_pars.append("--parallel")
+
+  if par["timestamps"]:
+    proc_pars.append("--timestamps")
+
+  ## Run pipeline
+  if not par["dryrun"]:
+    with tempfile.TemporaryDirectory(prefix="cwl-bd_rhapsody_wta-", dir=meta["temp_dir"]) as temp_dir:
+      cmd = ["cwl-runner"] + proc_pars + [cwl_file, os.path.basename(config_file)]
+
+      env = dict(os.environ)
+      env["TMPDIR"] = temp_dir
+
+      logger.info("> " + ' '.join(cmd))
+
+      try:
+        _ = subprocess.check_call(
+          cmd,
+          cwd=os.path.dirname(config_file),
+          env=env
+        )
+      finally:
+        shutil.rmtree(temp_dir)
```

Comment:
I think if we are going to use finally, we should not use a context manager. Additionally, I dont think rmtree will work here, because this is already used by the tempfile module itself: https://github.com/python/cpython/blob/main/Lib/tempfile.py#L862

Reply:
True, and yet sometimes the temp folder persists... :/

Reply:
Yes, I think we should use the `ignore_cleanup_errors` argument (with python3.10). This will try to remove as much of the files as possible, keeping the files that it could not delete. This might also give a hint as to which files are not being removed.